src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -66,21 +75,58 @@ public async Task<BundleResponse> Handle(BundleRequest bundleRequest, Cancellati
         {
             var bundleResource = bundleRequest.Bundle.ToPoco<Hl7.Fhir.Model.Bundle>();
 
-            if (bundleResource.Type != Hl7.Fhir.Model.Bundle.BundleType.Batch)
+            await FillRequestLists(bundleResource.Entry);
+
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Batch)
             {
-                throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.OnlyCertainBundleTypesSupported);
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
+                };
+
+                await ExecuteAllRequests(responseBundle);
+                return new BundleResponse(responseBundle.ToResourceElement());
             }
 
-            await FillRequestLists(bundleResource.Entry);
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Transaction)
+            {
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.TransactionResponse,
+                };
+
+                return await ExecuteTransactionForAllRequests(responseBundle);
+            }
+
+            throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.UnsupportedOperation);
+        }
+
+        private async Task<BundleResponse> ExecuteTransactionForAllRequests(Hl7.Fhir.Model.Bundle responseBundle)
+        {
+            BundleProcessingStatus bundleProcessingStatus = BundleProcessingStatus.SUCCEEDED;
 
-            var responseBundle = new Hl7.Fhir.Model.Bundle
+            try
             {
-                Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
-            };
+                try
+                {
+                    _transactionHandler.BeginTransactionScope();
+
+                    await ExecuteAllRequests(responseBundle);
 
-            await ExecuteAllRequests(responseBundle);
+                    _transactionHandler.CompleteTransactionScope();
+                }
+                finally
+                {
+                    _transactionHandler.Dispose();
+                }
+            }
+            catch (TransactionAbortedException ex)
+            {
+                bundleProcessingStatus = BundleProcessingStatus.FAILED;
```

Comment:
Could we just throw a new exception type here and handle the exception like the rest of the exceptions in the code base and return an operation outcome with a bad request status? Per the [spec](https://www.hl7.org/fhir/http.html#transaction-response) `For a failed transaction, the server returns a single OperationOutcome instead of a Bundle.`